### ISSUE TITLE
invalid needless_range_loop

### ISSUE THREAD

**Issue by madadam** [Manishearth/rust-clippy#1105]
Hello, 

I think it's an error to trigger the `needless_range_loop` lint in situation like this. The code cannot be transformed to what clippy suggests (to use an iterator), because the vector is mutably borrowed by the `mutate_all` function which prevents the iterator to borrow it too.

``` rust
fn main() {
    let mut numbers = vec![2, 4, 6, 8, 10];

    for i in 1..numbers.len() {
        numbers[i] += 1;
        mutate_all(&mut numbers);
    }
}

fn mutate_all(_numbers: &mut [i32]) {
    // ...
}
```

Clippy warning:

```
src/main.rs:7:5: 10:6 warning: the loop variable `i` is only used to index `numbers`., #[warn(needless_range_loop)] on by default
src/main.rs:7     for i in 1..numbers.len() {
                  ^
src/main.rs:7:9: 7:10 help: consider using an iterator
src/main.rs:      for <item> in numbers.iter().skip(1) {
src/main.rs:7:5: 10:6 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_range_loop
```

**Comment by pnkfelix** [comment]
Note that this also arises in (somewhat) more subtle cases like:

```rust
fn partition<T:PartialOrd+Send>(v: &mut [T]) -> usize {
    let pivot = v.len() - 1;
    let mut i = 0;
    for j in 0..pivot {
        if v[j] <= v[pivot] {
            v.swap(i, j);
            i += 1;
        }
    }
    v.swap(i, pivot);
    i
}
```

Here, the mutable borrow of `v` is hidden by an auto-borrowing method call. Make sure that if you fix this that you try to address that sort of case as well.

**Comment by devonhollowood** [comment]
I believe I just ran into this issue as well. I have the following code:
```rust
fn rotate_left<T: Clone>(slice: &mut [T], amount: usize) {
    let len = slice.len();
    for start_idx in 0..gcd(len, amount) {
        let temp = slice[start_idx].clone();
        let mut to = start_idx;
        let mut from = (to + amount) % len;
        while from != start_idx {
            slice[to] = slice[from].clone();
            to = from;
            from = (to + amount) % len;
        }
        slice[to] = temp;
    }
}
```

It generates the following `clippy` warning:
```
warning: the loop variable `start_idx` is used to index `slice`, #[warn(needless_range_loop)] on by default
   --> src/main.rs:158:5
    |
158 |     for start_idx in 0..gcd(len, amount) {
    |     ^
    |
help: consider using an iterator
    |     for (start_idx, <item>) in slice.iter().enumerate().take(gcd(len, amount)) {
    = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_range_loop
```

Now, when I change the code to this:
```rust
fn rotate_left<T: Clone>(slice: &mut [T], amount: usize) {
    let len = slice.len();
    for (start_idx, temp) in slice.iter().cloned().enumerate().take(gcd(len, amount)) {
        let mut to = start_idx;
        let mut from = (to + amount) % len;
        while from != start_idx {
            slice[to] = slice[from].clone();
            to = from;
            from = (to + amount) % len;
        }
        slice[to] = temp;
    }
}
```
I get the following borrow checker errors:
```
error[E0506]: cannot assign to `slice[..]` because it is borrowed
   --> src/main.rs:162:13
    |
158 |     for (start_idx, temp) in slice.iter().cloned().enumerate().take(gcd(len, amount)) {
    |                              ----- borrow of `slice[..]` occurs here
...
162 |             slice[to] = slice[from].clone();
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ assignment to borrowed `slice[..]` occurs here

error[E0506]: cannot assign to `slice[..]` because it is borrowed
   --> src/main.rs:166:9
    |
158 |     for (start_idx, temp) in slice.iter().cloned().enumerate().take(gcd(len, amount)) {
    |                              ----- borrow of `slice[..]` occurs here
...
166 |         slice[to] = temp;
    |         ^^^^^^^^^^^^^^^^ assignment to borrowed `slice[..]` occurs here
```

Incidentally, even if the borrow checking were to work out here, I'm not sure this lint is good advice in this case. The code it suggests is less clear than the original code in my opinion, because logically the cloning of `slice[start_idx]` into `temp` is separate from the indexing of `slice` via `start_idx`.

**Comment by pnkfelix** [comment]
How does one decide if the `needless_range_loop` is pulling its weight, given false positives such as the ones listed in this ticket?

**Comment by theemathas** [comment]
This is the only case I think `needless_range_loop` should catch.

    for some_variable in 0..some_container.len() {
        // Some code here.
        // When "some_container" appears, it always appears as "some_container[some_variable]".
    }
